webdriver-spec.html

```diff
@@ -5132,44 +5137,58 @@
+   <li><p>Let <var>keyboard</var> be a new <a>key input source</a>.
 
-  <li><p>Let <var>i</var> be 0.
+   <li><p>Let <var>undo actions</var> be an empty list.
 
-  <li><p>Let <var>length</var> be the length of <var>text</var>.
-
-  <li><p>While <var>i</var> &lt; <var>length</var>:
+   <li><p>For each <var>grapheme cluster</var> in <var>text</var>:
 
     <ol>
-      <li><p>Let <var>char</var> be the entry in <var>text</var> at
-        index <var>i</var>.
-
-      <li><p>Let <var>key to type</var> be the result of <a>processing keys</a>
-        with argument <var>key</var>.
-
-      <li><p>If the <var>key to type</var> is a <a>modifier key</a> then set the relevant
-        attribute on <a>Keyboard Event</a> to true for the rest of this loop.
-
-      <li><p>Generate <a data-lt=keyDown-Event>keyDown</a> <a>Keyboard Event</a>
-        with the relevant <a>keyboard event code</a> for <var>key to type</var>
-        against <var>active element</var>.
+     <li><p>Let <var>key to type</var> be the result of <a>processing
+      keys</a> with argument <var>grapheme cluster</var>.
+
+     <li><p>Let <var>keydown action</var> be an <a>action object</a>
+      constructed with
+      arguments <var>keyboard</var>'s <var>id</var>, <code>"key"</code>,
+      and <code>"keyDown"</code>.
+
+     <li><p>Set the <code>value</code> property of <var>keydown
+      action</var> to <var>key to type</var>.
+
+     <li><p><a>Dispatch a keyDown action</a> with
+      arguments <var>keydown action</var>
+      and <var>keyboard</var>'s <a>key input state</a>.
+
+     <li><p>Let <var>keyup action</var> be a copy of <var>keydown
+      action</var> with the <var>subtype</var> property changed
+      to <code>"keyUp"</code>.
+
+     <li><p>Run the substeps in the first matching statement:
+      <dl class="switch">
+       <dt><var>key to type</var> is a <a>modifier key</a>
+       <dd>Prepend <var>keyup action</var> to <var>cancel
+        actions</var>.
+
+        <dt><var>grapheme cluster</var> is the <a>null key</a>
+        <dd>For each <var>action</var> corresponding to an indexed
+          property in <var>undo actions</var> <a>dispatch a keyUp
+          action</a> with arguments <var>action</var>
+          and <var>keyboard</var>'s <a>key input
+          state</a>. Reset <var>undo actions</var> to an empty list.
+ 
+       <dt>Otherwise
+       <dd><a>Dispatch a keyUp action</a> with arguments <var>keyup
+        action</var> and <var>keyboard</var>'s <a>key input state</a>.
+
+      </dl>
+    </ol>
 
-      <li><p>Generate <a data-lt=keyPress-Event>keyPress</a> <a>Keyboard Event</a>
-        with the relevant <a>keyboard event code</a> for <var>key to type</var>
-        against <var>active element</var>.
+ <li><p>For each <var>action</var> corresponding to an indexed
+  property in <var>undo actions</var> <a>dispatch a keyUp action</a>
+  with arguments <var>action</var>
+  and <var>keyboard</var>'s <a>key input state</a>.
```

Comment:
This is repeated from above. I suggest we bring this out to a definition called, maybe, **clear the modifier keys state**. If you think it’s better to address this later, we can file an issue.